File: server/python_server_archive/grafana_utils.py
import datetime
from urllib.parse import parse_qs, urlparse

import requests


def parse_panel_url(panel_url):
    """Parse Grafana panel URL to extract host, UID, panel ID, and variables"""
    parsed = urlparse(panel_url)
    path_parts = parsed.path.strip("/").split("/")
    uid = path_parts[1] if len(path_parts) >= 2 else None
    query_params = parse_qs(parsed.query)
    panel_id_raw = query_params.get("viewPanel", [None])[0]
    panel_id = panel_id_raw.split("-")[-1] if panel_id_raw else None
    host = f"https://{parsed.netloc}"

    # Extract variables from URL
    variables = {}
    for key, values in query_params.items():
        if key.startswith("var-"):
            var_name = key[4:]  # Remove 'var-' prefix
            variables[var_name] = values  # Keep as list

    return host, uid, panel_id, variables


def get_dashboard_metadata(host, uid, token):
    """Fetch dashboard metadata from Grafana API"""
    dash_res = requests.get(
        f"{host}/api/dashboards/uid/{uid}",
        headers={"Authorization": f"Bearer {token}"},
    )
    dash_res.raise_for_status()
    return dash_res.json()


def apply_template_variables(data, variables):
    """Apply template variables to nested JSON data structure"""
    if not variables:
        return data

    def replace_variables_in_string(text):
        """Replace template variables in a string"""
        if not isinstance(text, str):
            return text

        result = text
        for var_name, var_value in variables.items():
            # Convert single values to list for consistent processing
            var_values = [var_value] if not isinstance(var_value, list) else var_value

            # Handle ${var-name} format (URL format)
            placeholder = f"${{var-{var_name}}}"
            if placeholder in result:
                result = result.replace(placeholder, var_values[0])

            # Handle ${name} format (dashboard JSON format)
            placeholder = f"${{{var_name}}}"
            if placeholder in result:
                # If multiple values, join with pipe for regex
                if len(var_values) > 1:
                    var_value_str = "|".join(var_values)
                    # For multi-value vars in Prometheus, often used in regex
                    result = result.replace(placeholder, f"({var_value_str})")
                else:
                    result = result.replace(placeholder, var_values[0])

            # Handle $name format (simple format)
            placeholder = f"${var_name}"
            if placeholder in result:
                # If multiple values, join with pipe for regex
                if len(var_values) > 1:
                    var_value_str = "|".join(var_values)
                    # For multi-value vars in Prometheus, often used in regex
                    result = result.replace(placeholder, f"({var_value_str})")
                else:
                    result = result.replace(placeholder, var_values[0])

        return result

    def process_data(obj):
        """Recursively process data structure"""
        if isinstance(obj, dict):
            return {key: process_data(value) for key, value in obj.items()}
        elif isinstance(obj, list):
            return [process_data(item) for item in obj]
        elif isinstance(obj, str):
            return replace_variables_in_string(obj)
        else:
            return obj

    return process_data(data)


def process_series_data(x_vals, y_vals, processed_targets, ref_id):
    """Process x and y values into chart data format"""
    # Check if x_vals contains timestamps or categorical data
    is_time_series = True
    if x_vals:
        first_val = x_vals[0]
        if isinstance(first_val, str) or (
            isinstance(first_val, (int, float)) and first_val < 1000000000
        ):
            is_time_series = False

    # Check target format preference
    target_format = "time_series"
    for target in processed_targets:
        if target.get("refId") == ref_id:
            target_format = target.get("format", "time_series")
            break

    if target_format == "table" or not is_time_series:
        # Categorical data (pie charts, bar charts, etc.)
        return [[str(x), y] for x, y in zip(x_vals, y_vals)]
    else:
        # Time series data
        return [
            [
                datetime.datetime.fromtimestamp(float(x) / 1000).isoformat(),
                float(y) if y is not None else 0,
            ]
            for x, y in zip(x_vals, y_vals)
            if x is not None and y is not None
        ]


def get_series_name_from_labels(frame, ref_id):
    """Extract series name from frame labels"""
    if "schema" in frame and "fields" in frame["schema"]:
        for field in frame["schema"]["fields"]:
            if field.get("type") == "number" and "labels" in field:
                labels = field["labels"]
                if "pod" in labels:
                    pod_name = labels["pod"]
                    if "resource-allocator" in pod_name:
                        parts = pod_name.split("-")
                        if len(parts) >= 3:
                            return f"{parts[-2][-4:]}-{parts[-1]}"
                        else:
                            return pod_name
                    else:
                        return pod_name
                elif "__name__" in labels:
                    return labels["__name__"]
                else:
                    # Use all meaningful labels to create name
                    label_parts = []
                    for k, v in labels.items():
                        if k not in ["__name__", "job", "instance"]:
                            label_parts.append(f"{k}={v}")
                    if label_parts:
                        return ", ".join(label_parts)
                break
    return ref_id


def query_grafana_panel(host, token, processed_targets, fr, to, panel_type):
    """Query Grafana API and process the response into structured data"""
    query_url = f"{host}/api/ds/query"

    # Use targets directly with variable substitution applied
    query_payload = {
        "queries": processed_targets,
        "from": str(fr),
        "to": str(to),
    }

    query_res = requests.post(
        query_url, json=query_payload, headers={"Authorization": f"Bearer {token}"}
    )
    response_data = query_res.json()
    query_res.raise_for_status()

    # Process the response data
    data_series = {}

    if "results" in response_data:
        # Process all series in the response
        for ref_id, result in response_data["results"].items():
            if "frames" in result and result["frames"]:
                # Each frame can be a separate series
                for frame in result["frames"]:
                    if "data" in frame and "values" in frame["data"]:
                        values = frame["data"]["values"]

                        # Check if this is a single value stat (like COUNT(*))
                        if len(values) == 1 and len(values[0]) == 1:
                            # Single stat value
                            stat_value = values[0][0]
                            data_series["stat_value"] = stat_value
                        elif len(values) >= 2:
                            x_vals = values[0]  # Time/category column

                            # Get field names from schema for series names
                            field_names = []
                            if "schema" in frame and "fields" in frame["schema"]:
                                field_names = [
                                    field.get("name", f"Series {i}")
                                    for i, field in enumerate(frame["schema"]["fields"])
                                ]

                            # Check if this is multi-column data (more than 2 columns)
                            if len(values) > 2 and len(field_names) > 2:
                                # Multi-column time series - each column after first is a series
                                for i in range(1, len(values)):
                                    y_vals = values[i]
                                    series_name = (
                                        field_names[i]
                                        if i < len(field_names)
                                        else f"Series {i}"
                                    )

                                    # Clean up series names
                                    series_name = (
                                        series_name.replace("_true", "")
                                        .replace("_", " ")
                                        .title()
                                    )

                                    frame_data = process_series_data(
                                        x_vals, y_vals, processed_targets, ref_id
                                    )
                                    if frame_data:
                                        data_series[series_name] = frame_data
                            else:
                                # Single series - process normally
                                y_vals = values[1]
                                series_name = get_series_name_from_labels(frame, ref_id)

                                # If no labels found, try to get name from target legendFormat
                                if series_name == ref_id:
                                    for target in processed_targets:
                                        if target.get("refId") == ref_id:
                                            legend_format = target.get(
                                                "legendFormat", ""
                                            )
                                            if (
                                                legend_format
                                                and "{{" not in legend_format
                                            ):
                                                series_name = legend_format
                                            elif target.get("expr"):
                                                expr = target.get("expr", "")[:40]
                                                series_name = (
                                                    f"{expr}..."
                                                    if len(expr) == 40
                                                    else expr
                                                )
                                            break

                                frame_data = process_series_data(
                                    x_vals, y_vals, processed_targets, ref_id
                                )
                                if frame_data:
                                    data_series[series_name] = frame_data
            elif "series" in result:
                # Time series format - each series is separate
                for series in result["series"]:
                    series_data = []
                    series_name = ref_id

                    # Try to get series name from tags/labels
                    if "tags" in series:
                        tags = series["tags"]
                        if "pod" in tags:
                            pod_name = tags["pod"]
                            if "resource-allocator" in pod_name:
                                parts = pod_name.split("-")
                                if len(parts) >= 3:
                                    series_name = f"{parts[-2][-4:]}-{parts[-1]}"
                                else:
                                    series_name = pod_name
                            else:
                                series_name = pod_name

                    for datapoint in series.get("datapoints", []):
                        if len(datapoint) >= 2 and datapoint[0] is not None:
                            value, timestamp = datapoint[0], datapoint[1]
                            series_data.append(
                                [
                                    datetime.datetime.fromtimestamp(
                                        timestamp / 1000
                                    ).isoformat(),
                                    value,
                                ]
                            )

                    # Add this series if it has data
                    if series_data:
                        data_series[series_name] = series_data

    # If no series data found, create empty single series
    if not data_series:
        data_series = {"No Data": []}

    return response_data, data_series

File: server/src/grafanaUtils.ts
import {
  DashboardMetadataException,
  GrafanaQueryException,
} from './exceptions';

export type GrafanaVariables = Record<string, string[]>;
export type GrafanaTarget = Record<string, any>;

export function parsePanelUrl(panelUrl: string): {
  host: string;
  uid: string | null;
  panelId: string | null;
  variables: GrafanaVariables;
} {
  const parsed = new URL(panelUrl);
  const pathParts = parsed.pathname
    .replace(/^\/+/g, '')
    .replace(/\/+$/g, '')
    .split('/');
  const uid = pathParts.length >= 2 ? pathParts[1] : null;
  const searchParams = new URLSearchParams(parsed.search);
  const panelIdRaw = searchParams.get('viewPanel');
  const panelId = panelIdRaw ? panelIdRaw.split('-').pop() || null : null;
  const host = `https://${parsed.host}`;

  const variables: GrafanaVariables = {};
  for (const key of Array.from(searchParams.keys())) {
    if (key.startsWith('var-')) {
      const varName = key.slice(4);
      variables[varName] = searchParams.getAll(key);
    }
  }

  return { host, uid, panelId, variables };
}

export async function getDashboardMetadata(
  host: string,
  uid: string,
  token: string,
): Promise<any> {
  const response = await fetch(`${host}/api/dashboards/uid/${uid}`, {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });

  const data = await response.json().catch(() => {
    throw new DashboardMetadataException('Invalid dashboard metadata response');
  });

  console.log(data);

  if (!response.ok) {
    throw new DashboardMetadataException(
      `Error fetching dashboard metadata: ${response.status} ${response.statusText}`,
    );
  }

  return data;
}

export function applyTemplateVariables(
  data: any,
  variables: GrafanaVariables,
): any {
  if (!variables || Object.keys(variables).length === 0) {
    return data;
  }

  const replaceVariablesInString = (text: any): any => {
    if (typeof text !== 'string') {
      return text;
    }

    let result = text;

    for (const [varName, varValue] of Object.entries(variables)) {
      const varValues = Array.isArray(varValue) ? varValue : [varValue];

      const patterns = [`\${var-${varName}}`, `\${${varName}}`, `$${varName}`];

      for (const pattern of patterns) {
        if (!result.includes(pattern)) {
          continue;
        }

        if (pattern === `\${${varName}}` || pattern === `$${varName}`) {
          const replacement =
            varValues.length > 1 ? `(${varValues.join('|')})` : varValues[0];
          result = result.split(pattern).join(replacement);
        } else {
          result = result.split(pattern).join(varValues[0]);
        }
      }
    }

    return result;
  };

  const processData = (obj: any): any => {
    if (Array.isArray(obj)) {
      return obj.map(processData);
    }

    if (obj && typeof obj === 'object') {
      return Object.fromEntries(
        Object.entries(obj).map(([key, value]) => [key, processData(value)]),
      );
    }

    return replaceVariablesInString(obj);
  };

  return processData(data);
}

export function processSeriesData(
  xVals: any[],
  yVals: any[],
  processedTargets: GrafanaTarget[],
  refId: string,
): any[] {
  let isTimeSeries = true;

  if (Array.isArray(xVals) && xVals.length > 0) {
    const firstVal = xVals[0];
    if (
      typeof firstVal === 'string' ||
      (typeof firstVal === 'number' && firstVal < 1000000000)
    ) {
      isTimeSeries = false;
    }
  }

  let targetFormat = 'time_series';
  for (const target of processedTargets) {
    if (target.refId === refId) {
      targetFormat = target.format ?? 'time_series';
      break;
    }
  }

  if (targetFormat === 'table' || !isTimeSeries) {
    return xVals.map((x, index) => [String(x), yVals[index]]);
  }

  return xVals
    .map((x, index) => {
      const y = yVals[index];
      if (x == null || y == null) {
        return null;
      }
      return [new Date(Number(x)).toISOString(), Number(y) || 0];
    })
    .filter(Boolean) as any[];
}

export function getSeriesNameFromLabels(frame: any, refId: string): string {
  if (frame?.schema?.fields) {
    for (const field of frame.schema.fields) {
      if (field.type === 'number' && field.labels) {
        const labels = field.labels;

        if (labels.pod) {
          const podName = labels.pod as string;
          if (podName.includes('resource-allocator')) {
            const parts = podName.split('-');
            if (parts.length >= 3) {
              return `${parts[parts.length - 2].slice(-4)}-${parts[parts.length - 1]}`;
            }
            return podName;
          }
          return podName;
        }

        if (labels.__name__) {
          return labels.__name__ as string;
        }

        const labelParts = Object.entries(labels)
          .filter(([key]) => !['__name__', 'job', 'instance'].includes(key))
          .map(([key, value]) => `${key}=${value}`);

        if (labelParts.length > 0) {
          return labelParts.join(', ');
        }
        break;
      }
    }
  }

  return refId;
}

export async function queryGrafanaPanel(
  host: string,
  token: string,
  processedTargets: GrafanaTarget[],
  fr: string,
  to: string,
  panelType: string,
): Promise<[any, any]> {
  const queryUrl = `${host}/api/ds/query`;
  const body = JSON.stringify({
    queries: processedTargets,
    from: String(fr),
    to: String(to),
  });

  const response = await fetch(queryUrl, {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${token}`,
      'Content-Type': 'application/json',
    },
    body,
  });

  const responseData = await response.json().catch(() => ({}));

  if (!response.ok) {
    throw new GrafanaQueryException(
      `Error querying Grafana panel: ${response.status} ${response.statusText} ${JSON.stringify(responseData)}`,
    );
  }

  const dataSeries: Record<string, any> = {};

  if (responseData.results) {
    const results = Object.entries(responseData.results) as [string, any][];
    for (const [refId, result] of results) {
      if (result?.frames?.length) {
        for (const frame of result.frames) {
          if (frame?.data?.values) {
            const values = frame.data.values as any[][];

            if (values.length === 1 && values[0].length === 1) {
              dataSeries.stat_value = values[0][0];
            } else if (values.length >= 2) {
              const xVals = values[0];
              const fieldNames = Array.isArray(frame?.schema?.fields)
                ? frame.schema.fields.map(
                    (field: any, index: number) =>
                      field.name ?? `Series ${index}`,
                  )
                : [];

              if (values.length > 2 && fieldNames.length > 2) {
                for (let i = 1; i < values.length; i += 1) {
                  const yVals = values[i];
                  const seriesName = fieldNames[i] ?? `Series ${i}`;
                  const frameData = processSeriesData(
                    xVals,
                    yVals,
                    processedTargets,
                    refId,
                  );
                  if (frameData.length > 0) {
                    dataSeries[seriesName] = frameData;
                  }
                }
              } else {
                const yVals = values[1];
                let seriesName = getSeriesNameFromLabels(frame, refId);

                if (seriesName === refId) {
                  for (const target of processedTargets) {
                    if (target.refId === refId) {
                      const legendFormat = target.legendFormat ?? '';
                      if (legendFormat && !legendFormat.includes('{{')) {
                        seriesName = legendFormat;
                      } else if (target.expr) {
                        const expr = String(target.expr).slice(0, 40);
                        seriesName = expr.length === 40 ? `${expr}...` : expr;
                      }
                      break;
                    }
                  }
                }

                const frameData = processSeriesData(
                  xVals,
                  yVals,
                  processedTargets,
                  refId,
                );
                if (frameData.length > 0) {
                  dataSeries[seriesName] = frameData;
                }
              }
            }
          }
        }
      } else if (result?.series?.length) {
        for (const series of result.series) {
          const seriesData: any[] = [];
          let seriesName = refId;

          if (series?.tags) {
            const tags = series.tags;
            if (tags.pod) {
              const podName = String(tags.pod);
              if (podName.includes('resource-allocator')) {
                const parts = podName.split('-');
                seriesName =
                  parts.length >= 3
                    ? `${parts[parts.length - 2].slice(-4)}-${parts[parts.length - 1]}`
                    : podName;
              } else {
                seriesName = podName;
              }
            }
          }

          for (const datapoint of series.datapoints ?? []) {
            if (
              Array.isArray(datapoint) &&
              datapoint.length >= 2 &&
              datapoint[0] != null
            ) {
              const [value, timestamp] = datapoint;
              seriesData.push([
                new Date(Number(timestamp)).toISOString(),
                value,
              ]);
            }
          }

          if (seriesData.length > 0) {
            dataSeries[seriesName] = seriesData;
          }
        }
      }
    }
  }

  if (Object.keys(dataSeries).length === 0) {
    dataSeries['No Data'] = [];
  }

  return [responseData, dataSeries];
}
